File: backend/app/schemas/email_crawler.py
"""
邮箱爬虫 Schema
"""
from pydantic import BaseModel, EmailStr, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class EmailTypeEnum(str, Enum):
    QQ = "qq"
    NETEASE_163 = "163"
    NETEASE_126 = "126"
    GMAIL = "gmail"
    OUTLOOK = "outlook"
    EXCHANGE = "exchange"
    OTHER = "other"


class ScanStatusEnum(str, Enum):
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class ImportStatusEnum(str, Enum):
    PENDING = "pending"
    IMPORTED = "imported"
    SKIPPED = "skipped"
    FAILED = "failed"


# ========== EmailAccount ==========

class EmailAccountBase(BaseModel):
    """邮箱账号基础"""
    email_address: str = Field(..., description="邮箱地址")
    email_type: EmailTypeEnum = Field(default=EmailTypeEnum.OTHER, description="邮箱类型")
    description: Optional[str] = Field(None, description="备注说明")
    imap_server: Optional[str] = Field(None, description="IMAP服务器（不填则使用预设）")
    imap_port: Optional[int] = Field(None, description="IMAP端口")
    use_ssl: bool = Field(default=True, description="是否使用SSL")
    username: str = Field(..., description="登录用户名")
    scan_folder: str = Field(default="INBOX", description="扫描文件夹")
    filter_sender: Optional[str] = Field(None, description="发件人过滤（多个用逗号分隔）")
    filter_subject: Optional[str] = Field(None, description="主题关键词过滤（多个用逗号分隔）")
    scan_days: int = Field(default=7, ge=1, le=365, description="扫描最近N天的邮件")


class EmailAccountCreate(EmailAccountBase):
    """创建邮箱账号"""
    password: str = Field(..., description="邮箱密码/授权码")


class EmailAccountUpdate(BaseModel):
    """更新邮箱账号"""
    email_address: Optional[str] = None
    email_type: Optional[EmailTypeEnum] = None
    description: Optional[str] = None
    imap_server: Optional[str] = None
    imap_port: Optional[int] = None
    use_ssl: Optional[bool] = None
    username: Optional[str] = None
    password: Optional[str] = Field(None, description="不填则不更新密码")
    scan_folder: Optional[str] = None
    filter_sender: Optional[str] = None
    filter_subject: Optional[str] = None
    scan_days: Optional[int] = Field(None, ge=1, le=365)
    is_active: Optional[bool] = None


class EmailAccountResponse(BaseModel):
    """邮箱账号响应"""
    id: int
    email_address: str
    email_type: EmailTypeEnum
    description: Optional[str]
    imap_server: str
    imap_port: int
    use_ssl: bool
    username: str
    scan_folder: str
    filter_sender: Optional[str]
    filter_subject: Optional[str]
    scan_days: int
    is_active: bool
    last_scan_at: Optional[datetime]
    last_scan_status: Optional[ScanStatusEnum]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class EmailAccountListResponse(BaseModel):
    """邮箱账号列表响应"""
    items: List[EmailAccountResponse]
    total: int


# ========== ScanLog ==========

class ScanLogResponse(BaseModel):
    """扫描日志响应"""
    id: int
    email_account_id: int
    scan_start: datetime
    scan_end: Optional[datetime]
    emails_found: int
    attachments_found: int
    parsed_success: int
    parsed_failed: int
    status: ScanStatusEnum
    error_message: Optional[str]
    
    # 额外信息
    email_address: Optional[str] = None  # 关联的邮箱地址
    
    class Config:
        from_attributes = True


class ScanLogListResponse(BaseModel):
    """扫描日志列表响应"""
    items: List[ScanLogResponse]
    total: int


# ========== PendingImport ==========

class PendingImportResponse(BaseModel):
    """待导入数据响应"""
    id: int
    scan_log_id: int
    email_subject: Optional[str]
    email_from: Optional[str]
    email_date: Optional[datetime]
    attachment_name: str
    attachment_size: Optional[int]
    product_code: Optional[str]
    product_name: Optional[str]
    product_id: Optional[int]
    nav_records_count: int
    parse_warnings: Optional[str]
    status: ImportStatusEnum
    import_result: Optional[str]
    imported_at: Optional[datetime]
    created_at: datetime
    
    # 预览数据（前几条净值）
    nav_preview: Optional[List[Dict[str, Any]]] = None
    
    class Config:
        from_attributes = True


class PendingImportListResponse(BaseModel):
    """待导入数据列表响应"""
    items: List[PendingImportResponse]
    total: int


class ImportConfirmRequest(BaseModel):
    """确认导入请求"""
    product_id: int = Field(..., description="目标产品ID")
    conflict_action: str = Field(default="skip", description="冲突处理方式：skip | overwrite")


class ImportResultResponse(BaseModel):
    """导入结果响应"""
    success: bool
    message: str
    imported: int = 0
    skipped: int = 0
    updated: int = 0


# ========== 其他 ==========

class TestConnectionResponse(BaseModel):
    """测试连接响应"""
    success: bool
    message: str


class ImapPresetResponse(BaseModel):
    """IMAP预设配置响应"""
    email_type: EmailTypeEnum
    server: str
    port: int
    ssl: bool

File: frontend/src/api/emailCrawler.ts
/**
 * 邮箱爬虫 API
 */
import request from "./request";

// ========== 类型定义 ==========

export type EmailType = 'qq' | '163' | '126' | 'gmail' | 'outlook' | 'exchange' | 'other'
export type ScanStatus = 'running' | 'completed' | 'failed'
export type ImportStatus = 'pending' | 'imported' | 'skipped' | 'failed'

export interface EmailAccount {
  id: number
  email_address: string
  email_type: EmailType
  description?: string
  imap_server: string
  imap_port: number
  use_ssl: boolean
  username: string
  scan_folder: string
  filter_sender?: string
  filter_subject?: string
  scan_days: number
  is_active: boolean
  last_scan_at?: string
  last_scan_status?: ScanStatus
  created_at: string
  updated_at: string
}

export interface EmailAccountCreate {
  email_address: string
  email_type: EmailType
  description?: string
  imap_server?: string
  imap_port?: number
  use_ssl?: boolean
  username: string
  password: string
  scan_folder?: string
  filter_sender?: string
  filter_subject?: string
  scan_days?: number
}

export interface EmailAccountUpdate {
  email_address?: string
  email_type?: EmailType
  description?: string
  imap_server?: string
  imap_port?: number
  use_ssl?: boolean
  username?: string
  password?: string
  scan_folder?: string
  filter_sender?: string
  filter_subject?: string
  scan_days?: number
  is_active?: boolean
}

export interface ScanLog {
  id: number
  email_account_id: number
  scan_start: string
  scan_end?: string
  emails_found: number
  attachments_found: number
  parsed_success: number
  parsed_failed: number
  status: ScanStatus
  error_message?: string
  email_address?: string
}

export interface NavPreview {
  nav_date: string
  unit_nav?: number
  cumulative_nav?: number
}

export interface PendingImport {
  id: number
  scan_log_id: number
  email_subject?: string
  email_from?: string
  email_date?: string
  attachment_name: string
  attachment_size?: number
  product_code?: string
  product_name?: string
  product_id?: number
  nav_records_count: number
  parse_warnings?: string
  status: ImportStatus
  import_result?: string
  imported_at?: string
  created_at: string
  nav_preview?: NavPreview[]
}

export interface ImapPreset {
  email_type: EmailType
  server: string
  port: number
  ssl: boolean
}

export interface ImportResult {
  success: boolean
  message: string
  imported: number
  skipped: number
  updated: number
}

// ========== API ==========

/**
 * 获取 IMAP 预设配置
 */
export function getImapPresets() {
  return request.get<ImapPreset[]>('/email-crawler/presets')
}

/**
 * 获取邮箱账号列表
 */
export function getEmailAccounts(params?: { is_active?: boolean }) {
  return request.get<{ items: EmailAccount[], total: number }>('/email-crawler/accounts', { params })
}

/**
 * 获取邮箱账号详情
 */
export function getEmailAccount(id: number) {
  return request.get<EmailAccount>(`/email-crawler/accounts/${id}`)
}

/**
 * 创建邮箱账号
 */
export function createEmailAccount(data: EmailAccountCreate) {
  return request.post<EmailAccount>('/email-crawler/accounts', data)
}

/**
 * 更新邮箱账号
 */
export function updateEmailAccount(id: number, data: EmailAccountUpdate) {
  return request.put<EmailAccount>(`/email-crawler/accounts/${id}`, data)
}

/**
 * 删除邮箱账号
 */
export function deleteEmailAccount(id: number) {
  return request.delete(`/email-crawler/accounts/${id}`)
}

/**
 * 测试邮箱连接
 */
export function testEmailConnection(id: number) {
  return request.post<{ success: boolean, message: string }>(`/email-crawler/accounts/${id}/test`)
}

/**
 * 扫描邮箱
 */
export function scanMailbox(id: number) {
  return request.post<ScanLog>(`/email-crawler/accounts/${id}/scan`)
}

/**
 * 获取扫描日志列表
 */
export function getScanLogs(params?: { account_id?: number, skip?: number, limit?: number }) {
  return request.get<{ items: ScanLog[], total: number }>('/email-crawler/logs', { params })
}

/**
 * 获取扫描日志详情
 */
export function getScanLog(id: number) {
  return request.get<ScanLog>(`/email-crawler/logs/${id}`)
}

/**
 * 获取待导入数据列表
 */
export function getPendingImports(params?: { status?: ImportStatus, scan_log_id?: number, skip?: number, limit?: number }) {
  return request.get<{ items: PendingImport[], total: number }>('/email-crawler/pending', { params })
}

/**
 * 获取待导入数据详情
 */
export function getPendingImport(id: number) {
  return request.get<PendingImport>(`/email-crawler/pending/${id}`)
}

/**
 * 确认导入
 */
export function confirmImport(id: number, data: { product_id: number, conflict_action: 'skip' | 'overwrite' }) {
  return request.post<ImportResult>(`/email-crawler/pending/${id}/import`, data)
}

/**
 * 跳过导入
 */
export function skipImport(id: number) {
  return request.post(`/email-crawler/pending/${id}/skip`)
}

// 邮箱类型显示名称
export const EMAIL_TYPE_LABELS: Record<EmailType, string> = {
  qq: 'QQ邮箱',
  '163': '网易163邮箱',
  '126': '网易126邮箱',
  gmail: 'Gmail (谷歌邮箱)',
  outlook: 'Outlook/Hotmail',
  exchange: 'Exchange/企业邮箱',
  other: '其他'
}

// 扫描状态显示
export const SCAN_STATUS_LABELS: Record<ScanStatus, string> = {
  running: '扫描中',
  completed: '已完成',
  failed: '失败'
}

// 导入状态显示
export const IMPORT_STATUS_LABELS: Record<ImportStatus, string> = {
  pending: '待处理',
  imported: '已导入',
  skipped: '已跳过',
  failed: '失败'
}
